rfcs/0031-devices-api.md

```diff
@@ -0,0 +1,363 @@
+- Feature Name: C Device API
+- Start Date: 02-08-2021
+- RFC PR: [apache/tvm-rfcs#31](https://github.com/apache/tvm-rfcs/pull/31)
+- GitHub Issue: [apache/tvm#0000](https://github.com/apache/tvm/issues/0000)
+
+
+# Summary
+[summary]: #summary
+This RFC aims to provide an API which can be used by the C runtime to abstract the variety of driver APIs for different platforms. This is specifically catering towards RTOS abstractions for embedded device drivers and aims to implement a subset of the overall Device API with supporting infrastructure to enable future expansion.
+
+# Motivation
+[motivation]: #motivation
+
+When using an accelerator, such as the [Arm&reg; Ethos&trade;-U](https://github.com/apache/tvm-rfcs/pull/11), an Embedded Real-Time Operating System (RTOS) will provide a device abstraction to access the device resource. When using these abstractions, TVM needs to understand how to interact with a device for a given platform.
+
+Taking the common example of a UART interface (imagining the accelerator is communicated to via this interface); in Zephyr, this would look similar to:
+
+```c
+#include <zephyr.h>
+#include <device.h>
+
+struct device *uart_dev = device_get_binding("USART0");
+
+char data[] = "Hello World!\r\n";
+uart_tx(uart_dev, data, sizeof(data), 100);
+```
+
+Whereas in CMSIS, this would look more similar to:
+
+```c
+ARM_DRIVER_USART* uart_dev = &Driver_USART0;
+uart_dev->Initialize(NULL);
+
+char data[] = "Hello World!\r\n";
+uart_dev->Send(data, sizeof(data)/sizeof(data[0]));
+```
+
+In this example, you can see the diversity of RTOS implementations for drivers and why it's required to provide a flexible abstraction to pass devices for micro targets.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## User App
+`tvm_device_<device>_t`s are implemented for each RTOS or platform required, these are included by the user who chooses as appropriate for their application. Notably, to avoid dynamic allocation, the user must provide the `tvm_device_<device>_t` struct and initialise it rather than it being created and setup for them in the API. This is augmented by named functions for each device, examples in the case of the "woofles" accelerator:
+
+```c
```

Comment:
can we include my life cycle picture and annotate with where each function is called? also can we document exactly why TVM calls each function and what it's expected to do (e.g. pre- and post-conditions)?

Reply:
I think this is included further down the RFC in the outline of an executor and explored further in future work? The hooks are intentionally kept simply named so they can be interpreted for a device, from my experience the `TVMDeviceWooflesActivate` could span from setting a pin high to sending a series of commands over a communications bus? The main thing I'm trying to communicate is that `Activate` is called before `Open` which is documented in the executor flow, this might not be the best way of demonstrating it though so please feel free to suggest an alternative.

Reply:
i think the reason i'm hoping for a life-cycle diagram is that the actions Activate/Open could take are quite varied. i agree the call scheme below is laid out in call order, but there's nothing explicit which states this so you just have to know that in advance. could you add something more explicit? i think the life-cycle diagram is the best thing as we can iterate on it as we add features here.